keps/sig-api-machinery/20190927-insecure-backend-proxy.md

```diff
@@ -0,0 +1,199 @@
+returning false log data.  If the user has chosen to accept this risk, we should allow it for the same reason we
+have an option for `--insecure-skip-tls-verify`.
+
+## Motivation
+
+On self-hosted clusters it is possible to end up in a state where a kubelet's serving certificate has expired so a kube-apiserver
+cannot verify the kubelet identity, *but* the kube-apiserver's client certificate is still valid so the kubelet can still
+verify the kube-apiserver.  In this condition, a cluster-admin may need to get pod logs to debug his cluster.
+
+### Goals
+
+1. Allow cluster-admins to get pod logs from kubelets with expired serving certificates.  This will include an API change
+and an addition argument to `kubectl log`
+
+### Non-Goals
+
+1. Allow any bidirectional traffic proxied to kubelets.  This may be a future objective, but is not in scope for the current KEP.
+
+## Proposal
+
+In [PodLogOptions](https://github.com/kubernetes/api/blob/d58b53da08f5430bb0f4e1154a73314e82b5b3aa/core/v1/types.go), 
+add a `InsecureSkipTLSVerifyBackend bool`
+```go
+// PodLogOptions is the query options for a Pod's logs REST call.
+type PodLogOptions struct {
+	// ... existing fields snipped
+	
+	// insecureSkipTLSVerifyBackend indicates that the apiserver should not confirm the validity of the of the
+	// serving certificate of the backend it is connecting to.  This will make the HTTPS connection between the apiserver
+	// and the backend insecure. This means that an attacker could intercept the logs request and provide false data.
+    // +optional
+    InsecureSkipTLSVerifyBackend bool `json:"insecureSkipTLSVerifyBackend,omitempty" protobuf:"varint,9,opt,name=insecureSkipTLSVerifyBackend"`
+}
+```
+The streamer for logs already prevents redirects, so an evil kubelet intercepting this traffic cannot redirect the kube-apiserver
+to use it's high powered credentials for a nefarious purpose.
+The `LocationStreamer` can take an additional argument and the `LogLocation`'s `NodeTransport` can actually produce a purpose
+built transport for insecure connections.
+
+To make this easier to use, we can add an `--insecure-skip-tls-verify-backend` flag to `kubectl log` which plumbs the option.
+This part of the KEP is a nice to have, since the kube-apiserver owns the backing capability which has general utility.
+
+### User Stories [optional]
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+This design is safe based on the following conditions:
+
+1. kube-apiservers only authenticate to kubelets using certificates.  See `--kubelet-client-certificate`.  Certificate based
+authentication does not send replayable credentials to backends.
+2. Clients must opt-in to the functionality and the documentation must include the impact in plain english.
+3. evil kubelets cannot trick kube-apiservers into using their credentials for any other purpose.
```

Comment:
is that really the case? E.g. if `--kubelet-client-certificate` matches the `--proxy-client-cert`, we can make the API server to identify as such to any target a malicious kubelet is proxying to.

Reply:
If the ending server terminates the TLS connection to honor the client, then the attacker (kubelet) did not get to choose the URL that is requested.  It's worth describing, but if you cannot choose the URLs, the "attack" is very limited in power.

Reply:
You are probably right. But I would like to have somebody take a look on the implications of `insecureSkipTLSVerifyBackend` who is deeply familiar with TLS and attack vectors.

Reply:
cc @kubernetes/sig-auth-pr-reviews

Reply:
Added clarification in the doc to explain that you can proxy a request for a URL getting logs for a pod.  To my knowledge, the proxier would not be able to rewrite that URL because it has to be terminated at the final destination.